rstudio/packrat
# Add environment masking to Packrat

### Intent

Mask certain environment variables from subprocesses launched by Packrat.

### Approach

There are two ways an environment variable can be added to the list of environment variables to mask:
- A predefined list of variables is built-in. This can be disabled by setting the option `packrat.mask.git.service.envvars` to `FALSE`.
- Users can define a character vector of additional variable names to mask in the option `packrat.masked.envvars`.

These are compiled in a function called `envvar_mask()`, which returns a named character vector of NAs with the names of variables to mask. This is used in the `R()` function that launches R subprocesses.

Other changes:
- A change was made to `system_check()` to add the argument `return_output = FALSE`, which captures and returns the output of the command, rather than returning `TRUE` upon success. This is used to facilitate tests that launch subprocesses and make assertions about the output.
- Renamed `is.named()`, which checks that all elements of a vector are named, to `all.named()`.

#### Open question

- Maybe the envvar mask should be loaded in `system_check()`. This is lower-level and can launch any process, whereas `R()` is for launching R processes. In practice, they are completely coupled.

### Automated Tests

Tests work one level out from the `envvar_mask()` function, launching an R process to run `R -e 'Sys.getenv()'` to examine the resulting environment. There are three tests:
- Test that the default list of variables is masked with the option unset
- Test that the default list of variables is ignored when the appropriate option is set
- Test that other variables can be masked

I wrote these tests before I moved the code to its own `envvar_mask()` function. It would be possible to test that inner code, but this outer test will fail if the inner code does not function.

### QA Notes

Install the test version of the package.

We're going to examine our environment with `Sys.getenv()`, and then run an R subprocess via Packrat and look at the environment seen by that subprocess. This will confirm that the new variable masking options are working correctly.

#### 1. Git service account variables are masked by default, but can be unmasked with an option.

In an R session, set an environment variable named `GITHUB_PAT`. The second line here should print out the result.

```r
Sys.setenv(GITHUB_PAT = "very_secret")
Sys.getenv("GITHUB_PAT")
```
```
[1] "very_secret"
```

Next, we'll see if it appears in a subprocess. This will print out a fair amount of text, including the R startup text. The result here is truncated. The important part is that the variable `very_secret` should not appear below the line where our function call is printed out.

```r
packrat:::R("-e 'Sys.getenv(\"GITHUB_PAT\")'", return_output = TRUE)
```
```
[20] "> Sys.getenv(\"GITHUB_PAT\")"                                 
[21] "[1] \"\""                                                     
```

Next, we'll turn off Git variable masking and see if the variable does appear.

```r
options(packrat.mask.git.service.envvars = FALSE)
packrat:::R("-e 'Sys.getenv(\"GITHUB_PAT\")'", return_output = TRUE)
```
```
[20] "> Sys.getenv(\"GITHUB_PAT\")"                                 
[21] "[1] \"very_secret\""                                          
```

#### 2. Arbitrary environment variables we provide can be masked with an option.

Set two variables and examine them locally.

```r
Sys.setenv(PLANT_BASED_PAT = "veggie_patty")
Sys.setenv(MEAT_EATERS_OPTION = "beef_patty")
Sys.getenv(c("PLANT_BASED_PAT", "MEAT_EATERS_OPTION"))
```
```
   PLANT_BASED_PAT MEAT_EATERS_OPTION 
    "veggie_patty"       "beef_patty" 
```

Check them in a subprocess.

```r
packrat:::R("-e 'Sys.getenv(c(\"PLANT_BASED_PAT\", \"MEAT_EATERS_OPTION\"))'", return_output = TRUE)
```
```
[20] "> Sys.getenv(c(\"PLANT_BASED_PAT\", \"MEAT_EATERS_OPTION\"))" 
[21] "   PLANT_BASED_PAT MEAT_EATERS_OPTION "                       
[22] "    \"veggie_patty\"       \"beef_patty\" "                   
```

Set new Packrat option and check that they are absent from the subprocess output.

```r
options(packrat.masked.envvars = c("PLANT_BASED_PAT", "MEAT_EATERS_OPTION"))
packrat:::R("-e 'Sys.getenv(c(\"PLANT_BASED_PAT\", \"MEAT_EATERS_OPTION\"))'", return_output = TRUE)
```
```
[20] "> Sys.getenv(c(\"PLANT_BASED_PAT\", \"MEAT_EATERS_OPTION\"))" 
[21] "   PLANT_BASED_PAT MEAT_EATERS_OPTION "                       
[22] "                \"\"                 \"\" "                   
```

#### 3. Verify that these options work together

Turn Git variable masking back on. Change the user mask option to only specify one of the variables. Confirm that all our variables still exist.

```r
options(packrat.mask.git.service.envvars = NULL)
options(packrat.masked.envvars = "PLANT_BASED_PAT")
Sys.getenv(c("GITHUB_PAT", "PLANT_BASED_PAT", "MEAT_EATERS_OPTION"))
```
```
        GITHUB_PAT    PLANT_BASED_PAT MEAT_EATERS_OPTION 
     "very_secret"     "veggie_patty"       "beef_patty" 
```

Check these variables in a subprocess. Only `MEAT_EATERS_OPTION` should be there.

```r
packrat:::R("-e 'Sys.getenv(c(\"GITHUB_PAT\", \"PLANT_BASED_PAT\", \"MEAT_EATERS_OPTION\"))'", return_output = TRUE)
```
```
[20] "> Sys.getenv(c(\"GITHUB_PAT\", \"PLANT_BASED_PAT\", \"MEAT_EATERS_OPTION\"))"
[21] "        GITHUB_PAT    PLANT_BASED_PAT MEAT_EATERS_OPTION "                   
[22] "                \"\"                 \"\"       \"beef_patty\" "             
```
